// Copyright (c) 2025 by OpenTier GmbH
// SPDX‑FileCopyrightText: 2025 OpenTier GmbH
// SPDX‑License‑Identifier: MIT
//
// This file is part of OpenTier.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

syntax = "proto3";

package robot;

import "vehicle_cloud_events.proto";

message RobotPacket {
  uint32 robot_id = 1;     // Target robot ID
  
  oneof payload {
    RobotCommand command = 2;
    RobotStatus status = 3;
  }
}

message RobotCommand {
  oneof command {
    Control control = 1;
    Kick kick = 2;
    Leds leds = 3;
    Beep beep = 4;
  }
}

message Control {
  float dx = 1;     // X velocity (m/s)
  float dy = 2;     // Y velocity (m/s)
  float dturn = 3;  // Rotation speed (rad/s)
}

message Kick {
  float power = 1;  // 0.0-1.0
}

message Leds {
  uint32 red = 1;    // 0-255
  uint32 green = 2;  // 0-255
  uint32 blue = 3;   // 0-255
}

message Beep {
  uint32 frequency = 1;  // Hz
  uint32 duration = 2;   // ms
}

message RobotStatus {
  float timestamp = 1;                   // Robot's internal time
  vehicle_cloud_events.BatteryEvent battery_info = 2; 
  bool is_running = 3;                   // Heartbeat active status
}
